#### 14 Oct 2024
#### Temas por revisar en este Notebook:

*   * Acuerdo con profesora guía:
*   * 1.Analizar las curvas del notebook Pruebas_Chile, usando valores absolutos y comparar
*   * 2.Validar los datos nuevamente y ver si explican bien la realidad

---


### Crear dataframe con la misma estructura que df, pero con los valores abolutos de las variables
*   1.Crear 'df_absoluto' con la misma estructura que 'df'
*   2.Crear lista con los nombres de los archivos en /notebooks/data/Annotated_Chile
*   3.Para cada elemento en los archivos de 'Annotated_Chile', llevarlo a un dataframe
*   4.Preprocesar los datos para asegurar que los valores de las etiquetas son nombres válidos de variables en el Manifesto Project dataset.
*
*   POBLAR DF_ABSOLUTOS
*   5.Hacer la matemática con las etiquetas
*   6.Crear 'df_absoluto' como copia de 'df'.
*   7.En 'df_absolutos' sólo dejar los registros de Chile y poner ceros en las variables 'per' 
*   8.En 'df_absolutos' agregar una columna 'num_frases'
*   8.En 'df_programas_Chile' contar las frases de cada programa y registrarlo en 'df_absoluto'  
*   6.Poblar 'df_absolutos', multiplicando los valores 'per' de 'df' por 'num_frases' 
*   7.Elegir un feature vector conveniente
*   8.Graficar con valores absolutos y comparar con gráficas hechas en base a % para el mismo feature vector
*   9.Concluir


#### Cargar librerías

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import Markdown, display
import glob

#### Cargar archivos

In [2]:
file_path = "data/MPDataset_MPDS2024a.csv"
df_Manifesto_project = pd.read_csv(file_path)
df_Manifesto_project.shape

/var/folders/v8/d2xyhw9n7pb9rqd7k0d1xfsc0000gn/T/ipykernel_1407/2392630412.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  df_Manifesto_project = pd.read_csv(file_path)


(5151, 175)

#### Copiar estructura

In [3]:
# df_absolutos = pd.DataFrame(columns=df_Manifesto_project.columns)
# df_absolutos.shape

#### Cargamos los programas *.csv en 'df_programas_Chile'

In [4]:
# Ruta a los archivos CSV
ruta_archivos = './data/Annotated_Chile/*.csv'

# Obtener la lista de archivos
archivos_csv = glob.glob(ruta_archivos)

# Verificar si se encontraron archivos
if not archivos_csv:
    print("No se encontraron archivos en la ruta especificada.")
else:
    print(f"Archivos encontrados: {archivos_csv}")

# Crear una lista para almacenar los DataFrames
dataframes = []

# Cargar los archivos al DataFrame
for archivo in archivos_csv:
    try:
        print(f"Leyendo archivo: {archivo}")
        df = pd.read_csv(
            archivo,
            encoding='utf-8',
            delimiter=',',
            skiprows=[1],  # Ignorar la primera fila después del encabezado
            names=['frase', 'frase_eng', 'cmp_code', 'eu_code'],  # Renombrar las columnas
            header=0,  # Ignorar los nombres originales de columnas
            keep_default_na=False,  # No interpretar valores como 'NA' como NaN
            na_values=[],  # No definir valores como NaN
            dtype={'cmp_code': str}  # Forzar la columna 'cmp_code' a tipo string
        )

        # Obtener el nombre del archivo sin ruta ni extensión
        nombre_archivo = archivo.split('/')[-1].replace('.csv', '')

        # Insertar la columna 'programa' como la primera columna
        df.insert(0, 'programa', nombre_archivo)

        # Agregar el DataFrame a la lista
        dataframes.append(df)
    except Exception as e:
        print(f"Error al leer {archivo}: {e}")

# Verificar si hay DataFrames para concatenar
if dataframes:
    df_programas_Chile = pd.concat(dataframes, ignore_index=True)
    print("DataFrame final:")
    print(df_programas_Chile.head())
else:
    print("No hay DataFrames para concatenar.")


Archivos encontrados: ['./data/Annotated_Chile/155021_200912.csv', './data/Annotated_Chile/155061_199312.csv', './data/Annotated_Chile/155601_200512.csv', './data/Annotated_Chile/155026_201711.csv', './data/Annotated_Chile/155980_201711.csv', './data/Annotated_Chile/155027_202111.csv', './data/Annotated_Chile/155620_198912.csv', './data/Annotated_Chile/155061_198912.csv', './data/Annotated_Chile/155501_201711.csv', './data/Annotated_Chile/155023_200912.csv', './data/Annotated_Chile/155201_201711.csv', './data/Annotated_Chile/155021_201311.csv', './data/Annotated_Chile/155021_199912.csv', './data/Annotated_Chile/155024_201311.csv', './data/Annotated_Chile/155021_201711.csv', './data/Annotated_Chile/155025_200512.csv', './data/Annotated_Chile/155061_202111.csv', './data/Annotated_Chile/155062_202111.csv', './data/Annotated_Chile/155061_200912.csv', './data/Annotated_Chile/155021_199312.csv', './data/Annotated_Chile/155025_200912.csv', './data/Annotated_Chile/155201_202111.csv', './data/A

### Preproceso a 'df_programas_Chile'
* Buscamos valores 'NaN', o Missing values, en la columna 'cmp_code' que es la que contiene los nombres de las etiquetas
* Eliminamos la columna 'eu_code' que no será utilizada

In [5]:
df_programas_Chile

programa                                              frase  \
0      155021_200912  La campaña presidencial es una buena oportunid...   
1      155021_200912                              Y así lo hemos hecho.   
2      155021_200912    Hemos vuelto a recorrer el país de norte a sur.   
3      155021_200912  Mis equipos de trabajo han recogido las inquie...   
4      155021_200912  Hemos recopilado historias de vida de miles de...   
...              ...                                                ...   
56935  155030_202111  Participaremos de modo incidente de organizaci...   
56936  155030_202111  5.4.5.Impulsaremos que la construcción de los ...   
56937  155030_202111  Propiciaremos que se desarrolle el carnet de i...   
56938  155030_202111  Propiciaremos que el Ministerio Público cuente...   
56939  155030_202111  Impulsaremos la creación de un Programa de Mem...   

                                               frase_eng cmp_code eu_code  
0      The presidential campaign is a good opportunit...    305.1      NA  
1                                        And so we have.    305.1      NA  
2      We have again toured the country from north to...    305.1      NA  
3      My work teams have gathered the concerns of pe...    305.1      NA  
4      We have collected life stories of thousands of...    305.1      NA  
...                                                  ...      ...     ...  
56935  We will participate incidentally with women's ...      503      NA  
56936  5.4.5.We will promote the incorporation of gen...        H      NA  
56937  We will promote the development of the gender-...      604      NA  
56938  We will promote that the Public Prosecutor's O...    605.1      NA  
56939  We will promote the creation of a Program of M...    201.2      NA  

[56940 rows x 5 columns]

#### Eliminamos la columna 'eu_code'. No la usaremos

In [6]:
df_programas_Chile = df_programas_Chile.drop(columns=['eu_code'])
df_programas_Chile.head()

programa                                              frase  \
0  155021_200912  La campaña presidencial es una buena oportunid...   
1  155021_200912                              Y así lo hemos hecho.   
2  155021_200912    Hemos vuelto a recorrer el país de norte a sur.   
3  155021_200912  Mis equipos de trabajo han recogido las inquie...   
4  155021_200912  Hemos recopilado historias de vida de miles de...   

                                           frase_eng cmp_code  
0  The presidential campaign is a good opportunit...    305.1  
1                                    And so we have.    305.1  
2  We have again toured the country from north to...    305.1  
3  My work teams have gathered the concerns of pe...    305.1  
4  We have collected life stories of thousands of...    305.1

##### Revisamos la existencia de valores 'no string' que aún puedan quedar en 'cmp_code'

In [7]:
# Count the occurrences of non-string values in the 'cmp_code' column
non_string_count = df_programas_Chile['cmp_code'].apply(lambda x: not isinstance(x, str)).sum()

print(f"Number of non-string values in 'cmp_code': {non_string_count}")

Number of non-string values in 'cmp_code': 0


##### Revisamos qué valores únicos quedan en 'cmp_code' para eliminar cualquier valor que no parezca un nombre de alguna de las variables

In [8]:
unique_cmp_code_values = df_programas_Chile['cmp_code'].unique()
unique_cmp_code_count = len(unique_cmp_code_values)

print(f"Unique values in 'cmp_code': {unique_cmp_code_values}")
print(f"Number of unique values in 'cmp_code': {unique_cmp_code_count}")

Unique values in 'cmp_code': ['305.1' '204' 'NA' '202.1' '303' '304' '301' '502' '305.6' '201.2'
 '606.1' '107' '202.4' '607.3' '503' '202.3' '106' '108' '104' '105' '410'
 '416.2' '701' '706' '506' '403' '402' '501' '414' '411' '703.1' '413'
 '504' '605.1' '605.2' '705' '505' '201.1' '305.4' '603' '203' '401' '405'
 '407' '109' '102' '305.2' '604' '704' '404' '101' '408' '606.2' 'H'
 '305.5' '602.1' '607.1' '415' '601.1' '103.1' '416.1' '607.2' '412' '409'
 '202.2' '602.2' '000' '702' '305.3' '103.2' '601.2' '406' '302' '110'
 '608.3' '507' '608.1' '608.2' '703.2']
Number of unique values in 'cmp_code': 79


##### Identificamos los registros en que 'cmp_code' == 'H' ó 'NA'. Dejaremos las ocurrencias de 'cmp_code' == '000'.

In [9]:
# Filter rows where cmp_code is 'NA' or 'H'
filtered_rows = df_programas_Chile[(df_programas_Chile['cmp_code'] == 'NA') | (df_programas_Chile['cmp_code'] == 'H')]

# Print the desired columns
for index, row in filtered_rows.iterrows():
    print(f"Row: {index}, Programa: {row['programa']}, Frase: {row['frase']}, CMP Code: {row['cmp_code']}")

Row: 67, Programa: 155021_200912, Frase: 1)  Impulsando una mejor democracia y un mejor Estado, CMP Code: NA
Row: 90, Programa: 155021_200912, Frase: HACIA UNA CONSTITUCIÓN DE TODOS, CMP Code: NA
Row: 107, Programa: 155021_200912, Frase: HACIA LA DIGNIFICACIÓN DE LA POLÍTICA Y EL SERVICIO PÚBLICO, CMP Code: NA
Row: 138, Programa: 155021_200912, Frase: HACIA UN PAÍS DESCENTRALIZADO Y CON GESTIÓN LOCAL, CMP Code: NA
Row: 155, Programa: 155021_200912, Frase: HACIA UNA REFORMA DEL ESTADO PARA FORTALECER EL ESTADO, CMP Code: NA
Row: 165, Programa: 155021_200912, Frase: HACIA MEJORES POLÍTICAS PÚBLICAS, CMP Code: NA
Row: 178, Programa: 155021_200912, Frase: HACIA UNA CULTURA DEL DIÁLOGO SOCIAL, CMP Code: NA
Row: 190, Programa: 155021_200912, Frase: HACIA UN PAÍS CON MEMORIA Y VERDAD, CMP Code: NA
Row: 208, Programa: 155021_200912, Frase: HACIA UN PAÍS DE VOLUNTARIOS PÚBLICOS, CMP Code: NA
Row: 224, Programa: 155021_200912, Frase: HACIA UNA MAYOR INTEGRACIÓN CON MUNDO MULTILATERAL, CMP Code: 

##### Vamos a eliminar los registros con 'cmp_code' == 'NA' ó 'H'. Los '000' los dejaremos porque pueden ser útiles después. 

In [10]:
# Count the number of rows to be removed
rows_to_remove = df_programas_Chile[(df_programas_Chile['cmp_code'] == 'NA') | (df_programas_Chile['cmp_code'] == 'H')].shape[0]

# Remove the rows
df_programas_Chile = df_programas_Chile[~((df_programas_Chile['cmp_code'] == 'NA') | (df_programas_Chile['cmp_code'] == 'H'))]

# Print the number of rows removed
print(f"Number of rows removed: {rows_to_remove}")

# Print the final number of rows
print(f"Final number of rows: {df_programas_Chile.shape[0]}")

Number of rows removed: 4095
Final number of rows: 52845


##### Reemplazamos los '.' en 'cmp_code' por un '_' para consistencia con el dataset del manifesto project

In [11]:
# Replace '.' with '_' in the 'cmp_code' column
df_programas_Chile['cmp_code'] = df_programas_Chile['cmp_code'].str.replace('.', '_', regex=False)

# Display the first few rows to verify the changes
df_programas_Chile.head()

programa                                              frase  \
0  155021_200912  La campaña presidencial es una buena oportunid...   
1  155021_200912                              Y así lo hemos hecho.   
2  155021_200912    Hemos vuelto a recorrer el país de norte a sur.   
3  155021_200912  Mis equipos de trabajo han recogido las inquie...   
4  155021_200912  Hemos recopilado historias de vida de miles de...   

                                           frase_eng cmp_code  
0  The presidential campaign is a good opportunit...    305_1  
1                                    And so we have.    305_1  
2  We have again toured the country from north to...    305_1  
3  My work teams have gathered the concerns of pe...    305_1  
4  We have collected life stories of thousands of...    305_1

##### Exportamos a un archivo excel

In [12]:
# Export df_programas_Chile to an Excel file
output_file_path = 'df_programas_Chile.xlsx'
df_programas_Chile.to_excel(output_file_path, index=False)
print(f"DataFrame exported to {output_file_path}")

DataFrame exported to df_programas_Chile.xlsx


##### Aseguramos que todos los valores en 'cmp_code' corresponde a nombres válidos de variables en el dataset de manifesto project. Estas variables están en el archivo 'Analytical_Data' 

In [13]:
# cargar Analytical_Data
df_AD = pd.read_excel('data/Analytical_Data.xlsx')
df_AD.head()

Dominio      Nombre_Dominio Variable_per  \
0  Domain_1  External Relations       per101   
1  Domain_1  External Relations       per102   
2  Domain_1  External Relations       per103   
3  Domain_1  External Relations       per104   
4  Domain_1  External Relations       per105   

                           Detalle_Dominio  
0  Foreign Special Relationships: Positive  
1  Foreign Special Relationships: Negative  
2                         Anti-Imperialism  
3                       Military: Positive  
4                       Military: Negative

In [14]:
# Iterar sobre los valores únicos de 'cmp_code' en 'df_programas_Chile'
for cmp_code in unique_cmp_code_values:
    # Anteponer 'per' al valor de 'cmp_code'
    per_code = f'per{cmp_code}'
    
    # Buscar el valor en la columna 'Variable_per' de 'df_AD'
    if per_code not in df_AD['Variable_per'].values:
        # Filtrar las filas correspondientes en 'df_programas_Chile'
        filtered_rows = df_programas_Chile[df_programas_Chile['cmp_code'] == cmp_code]
        
        # Imprimir la línea, programa y cmp_code
        for index, row in filtered_rows.iterrows():
            print(f"Línea: {index}, Programa: {row['programa']}, CMP Code: {row['cmp_code']}")

Línea: 6989, Programa: 155026_201711, CMP Code: 000
Línea: 7012, Programa: 155026_201711, CMP Code: 000
Línea: 7036, Programa: 155026_201711, CMP Code: 000
Línea: 7085, Programa: 155980_201711, CMP Code: 000
Línea: 7086, Programa: 155980_201711, CMP Code: 000
Línea: 7547, Programa: 155980_201711, CMP Code: 000
Línea: 7806, Programa: 155980_201711, CMP Code: 000
Línea: 9241, Programa: 155027_202111, CMP Code: 000
Línea: 10803, Programa: 155027_202111, CMP Code: 000
Línea: 16155, Programa: 155201_201711, CMP Code: 000
Línea: 28559, Programa: 155061_202111, CMP Code: 000
Línea: 28560, Programa: 155061_202111, CMP Code: 000
Línea: 28561, Programa: 155061_202111, CMP Code: 000
Línea: 28562, Programa: 155061_202111, CMP Code: 000
Línea: 28564, Programa: 155061_202111, CMP Code: 000
Línea: 28565, Programa: 155061_202111, CMP Code: 000
Línea: 28571, Programa: 155061_202111, CMP Code: 000
Línea: 28727, Programa: 155061_202111, CMP Code: 000
Línea: 28728, Programa: 155061_202111, CMP Code: 000
L

##### Sólo encontró los registros con '000'; es decir, todos los demás son nombres válidos de variables.

In [15]:
df_programas_Chile

programa                                              frase  \
0      155021_200912  La campaña presidencial es una buena oportunid...   
1      155021_200912                              Y así lo hemos hecho.   
2      155021_200912    Hemos vuelto a recorrer el país de norte a sur.   
3      155021_200912  Mis equipos de trabajo han recogido las inquie...   
4      155021_200912  Hemos recopilado historias de vida de miles de...   
...              ...                                                ...   
56934  155030_202111  Desarrollaremos la investigación de las necesi...   
56935  155030_202111  Participaremos de modo incidente de organizaci...   
56937  155030_202111  Propiciaremos que se desarrolle el carnet de i...   
56938  155030_202111  Propiciaremos que el Ministerio Público cuente...   
56939  155030_202111  Impulsaremos la creación de un Programa de Mem...   

                                               frase_eng cmp_code  
0      The presidential campaign is a good opportunit...    305_1  
1                                        And so we have.    305_1  
2      We have again toured the country from north to...    305_1  
3      My work teams have gathered the concerns of pe...    305_1  
4      We have collected life stories of thousands of...    305_1  
...                                                  ...      ...  
56934  We will develop research on the needs of careg...      504  
56935  We will participate incidentally with women's ...      503  
56937  We will promote the development of the gender-...      604  
56938  We will promote that the Public Prosecutor's O...    605_1  
56939  We will promote the creation of a Program of M...    201_2  

[52845 rows x 4 columns]

##### Revisión final de 'df_programas_Chile' buscando valores 'nan', 'NA', 'H', 'Missing values'

In [16]:
# Define the list of values to count
values_to_count = ['nan', 'NA', 'NAN', 'H', 'Missing values']

# Count the occurrences of each value in the 'cmp_code' column
occurrences = df_programas_Chile['cmp_code'].isin(values_to_count).sum()

print(f"Occurrences of specified values in 'cmp_code': {occurrences}")

Occurrences of specified values in 'cmp_code': 0


*   DF_ABSOLUTOS
*   5.Duplicamos 'df_Manifesto_Project' en 'df_absolutos'
*   6.xdc
*   7.En 'df_absolutos' sólo dejar los registros de Chile y poner ceros en las variables 'per' 
*   8.En 'df_absolutos' agregar una columna 'num_frases'
*   9.En 'df_programas_Chile' contar las frases de cada programa y registrarlo en 'df_absoluto'  
*   10.Poblar 'df_absolutos', multiplicando los valores 'per' de 'df' por 'num_frases' 
*   11.Elegir un feature vector conveniente
*   12.Graficar con valores absolutos y comparar con gráficas hechas en base a % para el mismo feature vector
*   13.Concluir

#### Creamos df_absolutos

In [17]:
df_absolutos = df_Manifesto_project.copy()

#### Dejamos sólo registros de Chile y agrega,ps la columna 'num_frases'

In [18]:
# Delete all records where 'countryname' is not 'Chile'
df_absolutos = df_absolutos[df_absolutos['countryname'] == 'Chile']

# Insert a new column named 'num_frases' with integer type
df_absolutos['num_frases'] = 0

# Display the first few rows to verify the changes
df_absolutos.head()

country countryname  oecdmember  eumember       edate    date   party  \
4895      155       Chile          10         0  14/12/1989  198912  155021   
4896      155       Chile          10         0  14/12/1989  198912  155061   
4897      155       Chile          10         0  14/12/1989  198912  155620   
4898      155       Chile          10         0  11/12/1993  199312  155021   
4899      155       Chile          10         0  11/12/1993  199312  155061   

                             partyname partyabbrev  parfam  ... per703_1  \
4895  Concert of Parties for Democracy         NaN      20  ...    2.165   
4896            Democracy and Progress         NaN      60  ...    1.171   
4897       Liberal-Socialists of Chile         NaN      60  ...    0.000   
4898  Concert of Parties for Democracy         NaN      20  ...    2.001   
4899        Union for Chile's Progress         NaN      60  ...    5.044   

      per703_2    rile  planeco  markeco  welfare  intpeace  datasetversion  \
4895       0.0 -15.152    4.205    3.958   12.369     1.855           2024a   
4896       0.0  21.405    0.167    8.863    7.692     1.672           2024a   
4897       0.0  -3.968    3.968    6.349   29.365     0.000           2024a   
4898       0.0 -10.493    5.368    3.758   13.860     0.537           2024a   
4899       0.0  25.347    1.513   15.385    7.818     0.252           2024a   

      id_perm  num_frases  
4895   LS71MA           0  
4896   SQXTCB           0  
4897   2IDQNL           0  
4898   6B4L5M           0  
4899   X97LN2           0  

[5 rows x 176 columns]

In [80]:
df_absolutos.shape

(30, 176)

#### En 'df_programas_Chile', contamos las frases por programa y lo registramos en 'num_frases' en df_absolutos'
#### Debemos homologar el nombre del programa en ambos dataframe y validar que los pares partido_fecha existen en ambos dataframe

In [ ]:
df_programas_Chile

In [21]:
# Extract unique values of 'programa' from 'df_programas_Chile'
unique_programas = df_programas_Chile['programa'].unique()

# Initialize a list to store warnings
warnings = []

# Iterate over unique values of 'programa'
for programa in unique_programas:
    # Extract the leftmost 6 characters and rightmost 6 characters
    left_6 = programa[:6]
    right_6 = programa[-6:]
    
    # Validate if 'party' and 'date' exist in 'df_absolutos'
    if not ((df_absolutos['party'] == int(left_6)) & (df_absolutos['date'] == int(right_6))).any():
        warnings.append(f"Warning: No matching row found for 'party' == {left_6} and 'date' == {right_6}")
    else:
        # Count the number of occurrences of 'programa' in 'df_programas_Chile'
        num_frases = df_programas_Chile[df_programas_Chile['programa'] == programa].shape[0]
        
        # Update 'num_frases' in 'df_absolutos'
        df_absolutos.loc[(df_absolutos['party'] == int(left_6)) & (df_absolutos['date'] == int(right_6)), 'num_frases'] = num_frases

# Print warnings if any
for warning in warnings:
    print(warning)

In [24]:
df_absolutos.head()

country countryname  oecdmember  eumember       edate    date   party  \
4895      155       Chile          10         0  14/12/1989  198912  155021   
4896      155       Chile          10         0  14/12/1989  198912  155061   
4897      155       Chile          10         0  14/12/1989  198912  155620   
4898      155       Chile          10         0  11/12/1993  199312  155021   
4899      155       Chile          10         0  11/12/1993  199312  155061   

                             partyname partyabbrev  parfam  ... per703_1  \
4895  Concert of Parties for Democracy         NaN      20  ...    2.165   
4896            Democracy and Progress         NaN      60  ...    1.171   
4897       Liberal-Socialists of Chile         NaN      60  ...    0.000   
4898  Concert of Parties for Democracy         NaN      20  ...    2.001   
4899        Union for Chile's Progress         NaN      60  ...    5.044   

      per703_2    rile  planeco  markeco  welfare  intpeace  datasetversion  \
4895       0.0 -15.152    4.205    3.958   12.369     1.855           2024a   
4896       0.0  21.405    0.167    8.863    7.692     1.672           2024a   
4897       0.0  -3.968    3.968    6.349   29.365     0.000           2024a   
4898       0.0 -10.493    5.368    3.758   13.860     0.537           2024a   
4899       0.0  25.347    1.513   15.385    7.818     0.252           2024a   

      id_perm  num_frases  
4895   LS71MA        1616  
4896   SQXTCB         597  
4897   2IDQNL         125  
4898   6B4L5M        2048  
4899   X97LN2         792  

[5 rows x 176 columns]

#### En 'df_absolutos', cada columna 'per' que representa una variable que nos interesa, la multiplicamos por 'num_frases'

In [28]:
# Create a list of column names that begin with 'per' and have a digit or an underscore in position 4
variables_per = [col for col in df_absolutos.columns if col.startswith('per') and (col[3].isdigit() or col[3] == '_')]

# Display the list
print(f"Total variables per: {len(variables_per)}")
# variables_per


Total variables per: 142


In [29]:
# Iterate over each row in df_absolutos
for index, row in df_absolutos.iterrows():
    # Multiply each value in variables_per by num_frases and format as integer
    for col in variables_per:
        df_absolutos.at[index, col] = int(row[col] * row['num_frases'])

# Display the first few rows to verify the changes
df_absolutos.head()

country countryname  oecdmember  eumember       edate    date   party  \
4895      155       Chile          10         0  14/12/1989  198912  155021   
4896      155       Chile          10         0  14/12/1989  198912  155061   
4897      155       Chile          10         0  14/12/1989  198912  155620   
4898      155       Chile          10         0  11/12/1993  199312  155021   
4899      155       Chile          10         0  11/12/1993  199312  155061   

                             partyname partyabbrev  parfam  ... per703_1  \
4895  Concert of Parties for Democracy         NaN      20  ...   3498.0   
4896            Democracy and Progress         NaN      60  ...    699.0   
4897       Liberal-Socialists of Chile         NaN      60  ...      0.0   
4898  Concert of Parties for Democracy         NaN      20  ...   4098.0   
4899        Union for Chile's Progress         NaN      60  ...   3994.0   

      per703_2    rile  planeco  markeco  welfare  intpeace  datasetversion  \
4895       0.0 -15.152    4.205    3.958   12.369     1.855           2024a   
4896       0.0  21.405    0.167    8.863    7.692     1.672           2024a   
4897       0.0  -3.968    3.968    6.349   29.365     0.000           2024a   
4898       0.0 -10.493    5.368    3.758   13.860     0.537           2024a   
4899       0.0  25.347    1.513   15.385    7.818     0.252           2024a   

      id_perm  num_frases  
4895   LS71MA        1616  
4896   SQXTCB         597  
4897   2IDQNL         125  
4898   6B4L5M        2048  
4899   X97LN2         792  

[5 rows x 176 columns]

In [ ]:
# Display the first few rows of the DataFrame
df.head()

In [ ]:
# Corrige nombre de Se
# Count the number of changes
count_changes = df.loc[df['candidatename'].notna() & df['candidatename'].str.contains('Piñera'), 'candidatename'].count()

# Replace the values
df.loc[df['candidatename'].notna() & df['candidatename'].str.contains('Piñera'), 'candidatename'] = 'Sebastián Piñera Echeñique'

# Print the number of changes
print(f"Number of changes: {count_changes}")

#### Crear campo 'agno' para registrar el año del programa de gobierno

In [4]:
# Insert column 'agno' into the dataframe, of type int, to the right of column 'date'
df.insert(df.columns.get_loc('date') + 1, 'agno', df['edate'].str[-4:].astype(int))

# List unique values of new column 'agno'
unique_agno_values = df['agno'].unique()
#print(unique_agno_values)

In [ ]:
# cargar Analytical_Data
df_AD = pd.read_excel('data/Analytical_Data.xlsx')
df_AD.head()


In [6]:
# Create a list of column names that begin with 'per' and have a digit in position 4
per_columns = ['countryname', 'candidatename', 'agno', 'presvote'] + [col for col in df.columns if col.startswith('per') and col[3].isdigit()]
#per_columns

#### PROMEDIO PONDERADO: en 'df' filtramos registros de Chile, separamos por año de elección y calculamos valores representativos basados en el promedio ponderado por el % obtenido en la elección presidencial (presvote)

In [ ]:
# Filter df by 'countryname' == 'Chile'
df_Chile = df[df['countryname'] == 'Chile']

df_Chile = df_Chile.sort_values(by='agno')

# Create a list of column names that begin with 'per' and have a digit in position 4
per_columns_4digit = [col for col in df.columns if col.startswith('per') and col[3].isdigit()]

# Group by column 'agno' and calculate the weighted average of each column in 'per_columns_4digit' using column 'presvote' as weights
df_Chile_ponderado = df_Chile.groupby('agno').apply(
    lambda x: pd.Series({col: np.average(x[col], weights=x['presvote']) for col in per_columns_4digit})
).reset_index()

# Display the results
df_Chile_ponderado

#### INCLUSIÓN Y MINORÍAS (503, 705, 201_2) 

In [ ]:
# Define variables for the columns
tema = "INCLUSIÓN y MINORÍAS"
col_1 = 'per503'
col_2 = 'per705'
col_3 = 'per201_2'

# Select columns from df_Chile_ponderado
df_Chile_inclymin = df_Chile_ponderado[['agno', col_1, col_2, col_3]]

# Add a new column 'Total' with the sum of columns
df_Chile_inclymin['Total'] = df_Chile_inclymin[col_1] + df_Chile_inclymin[col_2] + df_Chile_inclymin[col_3]

# Get the column names excluding 'agno'
columns_to_lookup = df_Chile_inclymin.columns[df_Chile_inclymin.columns != 'agno']
detalle_dominio_dict = {}

# Loop through the columns and get the corresponding 'Detalle_Dominio' from 'df_AD'
for col in columns_to_lookup:
    detalle_dominio = df_AD.loc[df_AD['Variable_per'] == col, 'Detalle_Dominio'].values
    if len(detalle_dominio) > 0:
        detalle_dominio_dict[col] = detalle_dominio[0]

# Update the plot labels with details from the dictionary
label_col_1 = f"{col_1} ({detalle_dominio_dict.get(col_1, 'No Info')})"
label_col_2 = f"{col_2} ({detalle_dominio_dict.get(col_2, 'No Info')})"
label_col_3 = f"{col_3} ({detalle_dominio_dict.get(col_3, 'No Info')})"
label_total = f"Total ({col_1} + {col_2} + {col_3})"

texto_md = """<div style="display: flex; justify-content: space-between;">
  <img src="../assets/per503.png" alt="" width="450">
  <img src="../assets/per705.png" alt="" width="450"></div>
</div>
<img src="../assets/per201_2.png" alt="" width="500">"""
display(Markdown(texto_md))

# Plot columns
plt.figure(figsize=(10, 6))
plt.plot(df_Chile_inclymin['agno'], df_Chile_inclymin[col_1], label=label_col_1)
plt.plot(df_Chile_inclymin['agno'], df_Chile_inclymin[col_2], label=label_col_2)
plt.plot(df_Chile_inclymin['agno'], df_Chile_inclymin[col_3], label=label_col_3)
plt.plot(df_Chile_inclymin['agno'], df_Chile_inclymin['Total'], label=label_total)

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title(f'PAÍS: CHILE\n{tema}')
plt.legend()
plt.grid(True)
plt.show()


#### Comentario a Inclusión y Minorías:
* La línea del Total muestra que el aspecto "Inclusión y Minorías" ha ido disminuyendo porcentualmente en las menciones de los programas de gobierno entre 1989 y 2021. Sin embargo es el segundo más importante de los aspectos en análisis, lo cual se puede apreciar en el gráfico resumen.

#### ORDEN y CORRUPCIÓN (605_1, -605_2, per304)

In [ ]:
# Define variables for the columns
tema = "ORDEN Y CORRUPCIÓN"
col_1 = 'per605_1'
col_2 = 'per605_2'
col_3 = 'per304'

# Select columns from df_Chile_ponderado
df_Chile_orden = df_Chile_ponderado[['agno', col_1, col_2, col_3]]

# Add a new column 'Total' with the sum of columns
df_Chile_orden['Total'] = df_Chile_orden[col_1] - df_Chile_orden[col_2] + df_Chile_orden[col_3]

# Get the column names excluding 'agno'
columns_to_lookup = df_Chile_orden.columns[df_Chile_orden.columns != 'agno']
detalle_dominio_dict = {}

# Loop through the columns and get the corresponding 'Detalle_Dominio' from 'df_AD'
for col in columns_to_lookup:
    detalle_dominio = df_AD.loc[df_AD['Variable_per'] == col, 'Detalle_Dominio'].values
    if len(detalle_dominio) > 0:
        detalle_dominio_dict[col] = detalle_dominio[0]

# Update the plot labels with details from the dictionary
label_col_1 = f"{col_1} ({detalle_dominio_dict.get(col_1, 'No Info')})"
label_col_2 = f"{col_2} ({detalle_dominio_dict.get(col_2, 'No Info')})"
label_col_3 = f"{col_3} ({detalle_dominio_dict.get(col_3, 'No Info')})"
label_total = f"Total ({col_1} - {col_2} + {col_3})"

texto_md = """<div style="display: flex; justify-content: space-between;">
  <img src="../assets/per605_1-per605_2.png" alt="" width="450"></div>
  <img src="../assets/per304.png" alt="" width="450">
</div>

"""
display(Markdown(texto_md))

# Plot columns
plt.figure(figsize=(10, 6))
plt.plot(df_Chile_orden['agno'], df_Chile_orden[col_1], label=label_col_1)
plt.plot(df_Chile_orden['agno'], df_Chile_orden[col_2], label=label_col_2)
plt.plot(df_Chile_orden['agno'], df_Chile_orden[col_3], label=label_col_3)
plt.plot(df_Chile_orden['agno'], df_Chile_orden['Total'], label=label_total)

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title(f'PAÍS: CHILE\n{tema} ({col_1} - {col_2} + {col_3})')
plt.legend()
plt.grid(True)
plt.show()


#### Comentario a ORDEN Y CORRUPCIÓN
* Se puede apreciar claramente la importancia que ha ido adquiriendo este aspecto en Chile, variando desde 1,1 en 1989 a 5,2 en 2021 (4.8 veces). Esto ha sido impulsado por un fuerte aumento de las menciones a dotar de mayores recursos a las policías y mayor rigurosidad en los tribunales. Debe mencionarse la disminución de las menciones a la corrupción política y al abuso de este poder.  

#### MEDIO AMBIENTE Y SOSTENIBILIDAD (501, 416_2)

In [ ]:
# Define variables for the columns
tema = "MEDIO AMBIENTE Y SOSTENIBILIDAD"
col_1 = 'per501'
col_2 = 'per416_2'

# Select columns from df_Chile_ponderado
df_Chile_medioambiente = df_Chile_ponderado[['agno', col_1, col_2]]

# Add a new column 'Total' with the sum of columns
df_Chile_medioambiente['Total'] = df_Chile_medioambiente[col_1] + df_Chile_medioambiente[col_2]

# Get the column names excluding 'agno'
columns_to_lookup = df_Chile_medioambiente.columns[df_Chile_medioambiente.columns != 'agno']
detalle_dominio_dict = {}

# Loop through the columns and get the corresponding 'Detalle_Dominio' from 'df_AD'
for col in columns_to_lookup:
    detalle_dominio = df_AD.loc[df_AD['Variable_per'] == col, 'Detalle_Dominio'].values
    if len(detalle_dominio) > 0:
        detalle_dominio_dict[col] = detalle_dominio[0]

# Update the plot labels with details from the dictionary
label_col_1 = f"{col_1} ({detalle_dominio_dict.get(col_1, 'No Info')})"
label_col_2 = f"{col_2} ({detalle_dominio_dict.get(col_2, 'No Info')})"
label_total = f"Total ({col_1} + {col_2})"

texto_md = """<div style="display: flex; justify-content: space-between;">
  <img src="../assets/per501.png" alt="" width="550"></div>
  <img src="../assets/per416_2.png" alt="" width="450">
</div>
"""
display(Markdown(texto_md))

# Plot columns
plt.figure(figsize=(10, 6))
plt.plot(df_Chile_medioambiente['agno'], df_Chile_medioambiente[col_1], label=label_col_1)
plt.plot(df_Chile_medioambiente['agno'], df_Chile_medioambiente[col_2], label=label_col_2)
plt.plot(df_Chile_medioambiente['agno'], df_Chile_medioambiente['Total'], label=label_total)

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title(f'PAÍS: CHILE\n{tema} ({col_1} + {col_2})')
plt.legend()
plt.grid(True)
plt.show()


#### Comentario a MEDIO AMBIENTE Y SOSTENIBILIDAD
* Este es el aspecto que evoluciona más fuerte de todos, y en todos sus componentes. 

#### PUEBLOS ORIGINARIOS (607_3, 608_3)

In [ ]:
# Define variables for the columns
tema = "PUEBLOS ORIGINARIOS"
col_1 = 'per607_3'
col_2 = 'per608_3'

# Select columns from df_Chile_ponderado
df_Chile_poriginarios = df_Chile_ponderado[['agno', col_1, col_2]]

# Add a new column 'Total' with the sum of columns
df_Chile_poriginarios['Total'] = df_Chile_poriginarios[col_1] - df_Chile_poriginarios[col_2]

# Get the column names excluding 'agno'
columns_to_lookup = df_Chile_poriginarios.columns[df_Chile_poriginarios.columns != 'agno']
detalle_dominio_dict = {}

# Loop through the columns and get the corresponding 'Detalle_Dominio' from 'df_AD'
for col in columns_to_lookup:
    detalle_dominio = df_AD.loc[df_AD['Variable_per'] == col, 'Detalle_Dominio'].values
    if len(detalle_dominio) > 0:
        detalle_dominio_dict[col] = detalle_dominio[0]

# Update the plot labels with details from the dictionary
label_col_1 = f"{col_1} ({detalle_dominio_dict.get(col_1, 'No Info')})"
label_col_2 = f"{col_2} ({detalle_dominio_dict.get(col_2, 'No Info')})"
label_total = f"Total ({col_1} - {col_2})"

texto_md = """<div style="display: flex; justify-content: space-between;">
  <img src="../assets/per607_3.png" alt="" width="550"></div>
  <img src="../assets/per608_3.png" alt="" width="550">
</div>
"""
display(Markdown(texto_md))

# Plot columns
plt.figure(figsize=(10, 6))
plt.plot(df_Chile_poriginarios['agno'], df_Chile_poriginarios[col_1], label=label_col_1)
plt.plot(df_Chile_poriginarios['agno'], df_Chile_poriginarios[col_2], label=label_col_2)
plt.plot(df_Chile_poriginarios['agno'], df_Chile_poriginarios['Total'], label=label_total)

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title(f'PAÍS: CHILE\n{tema} ({col_1} - {col_2})')
plt.legend()
plt.grid(True)
plt.show()


#### Comentario a PUEBLOS ORIGINARIOS
* Luego de tener un crecimiento fuerte y sostenido entre 1989 y 2008, PUEBLOS ORIGINARIOS ha ido perdiendo fuertemente participación en las menciones en los programas de gobierno. 

#### DERECHOS HUMANOS (per201_1, per201_2)

In [ ]:
# Define variables for the columns
tema = "DERECHOS HUMANOS"
col_1 = 'per201_1'
col_2 = 'per201_2'

# Select columns from df_Chile_ponderado
df_Chile_hr = df_Chile_ponderado[['agno', col_1, col_2]]

# Add a new column 'Total' with the sum of columns
df_Chile_hr['Total'] = df_Chile_hr[col_1] + df_Chile_hr[col_2]

# Get the column names excluding 'agno'
columns_to_lookup = df_Chile_hr.columns[df_Chile_hr.columns != 'agno']
detalle_dominio_dict = {}

# Loop through the columns and get the corresponding 'Detalle_Dominio' from 'df_AD'
for col in columns_to_lookup:
    detalle_dominio = df_AD.loc[df_AD['Variable_per'] == col, 'Detalle_Dominio'].values
    if len(detalle_dominio) > 0:
        detalle_dominio_dict[col] = detalle_dominio[0]

# Update the plot labels with details from the dictionary
label_col_1 = f"{col_1} ({detalle_dominio_dict.get(col_1, 'No Info')})"
label_col_2 = f"{col_2} ({detalle_dominio_dict.get(col_2, 'No Info')})"
label_total = f"Total ({col_1} + {col_2})"

texto_md = """<div style="display: flex; justify-content: space-between;">
  <img src="../assets/per201_1_2.png" alt="" width="550"></div>
</div>
"""
display(Markdown(texto_md))

# Plot columns
plt.figure(figsize=(10, 6))
plt.plot(df_Chile_hr['agno'], df_Chile_hr[col_1], label=label_col_1)
plt.plot(df_Chile_hr['agno'], df_Chile_hr[col_2], label=label_col_2)
plt.plot(df_Chile_hr['agno'], df_Chile_hr['Total'], label=label_total)

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title(f'PAÍS: CHILE\n{tema} ({col_1} + {col_2})')
plt.legend()
plt.grid(True)
plt.show()


#### Comentario a DERECHOS HUMANOS:
* Este aspecto ha ido perdiendo participación desde 1989 a 2021, pero crece en forma sostenida desde 2005.

#### INMIGRACIÓN (602_2, 607_1)

In [ ]:
# Define variables for the columns
tema = "INMIGRACIÓN"
col_1 = 'per602_2'
col_2 = 'per607_1'
#col_3 = 'per201_2'

# Select columns from df_Chile_ponderado
df_Chile_inmigracion = df_Chile_ponderado[['agno', col_1, col_2]] #col_3

# Add a new column 'Total' with the sum of columns
df_Chile_inmigracion['Total'] = df_Chile_inmigracion[col_1] + df_Chile_inmigracion[col_2] #+ df_Chile_inmigracion[col_3]

# Get the column names excluding 'agno'
columns_to_lookup = df_Chile_inmigracion.columns[df_Chile_inmigracion.columns != 'agno']
detalle_dominio_dict = {}

# Loop through the columns and get the corresponding 'Detalle_Dominio' from 'df_AD'
for col in columns_to_lookup:
    detalle_dominio = df_AD.loc[df_AD['Variable_per'] == col, 'Detalle_Dominio'].values
    if len(detalle_dominio) > 0:
        detalle_dominio_dict[col] = detalle_dominio[0]

# Update the plot labels with details from the dictionary
label_col_1 = f"{col_1} ({detalle_dominio_dict.get(col_1, 'No Info')})"
label_col_2 = f"{col_2} ({detalle_dominio_dict.get(col_2, 'No Info')})"
#label_col_3 = f"{col_3} ({detalle_dominio_dict.get(col_3, 'No Info')})"
label_total = f"Total ({col_1} + {col_2})" # + {col_3})"

texto_md = """<div style="display: flex; justify-content: space-between;">
  <img src="../assets/per602_2.png" alt="" width="550"></div>
  <img src="../assets/per607_1.png" alt="" width="550">
</div>
"""
display(Markdown(texto_md))

# Plot columns
plt.figure(figsize=(10, 6))
plt.plot(df_Chile_inmigracion['agno'], df_Chile_inmigracion[col_1], label=label_col_1)
plt.plot(df_Chile_inmigracion['agno'], df_Chile_inmigracion[col_2], label=label_col_2)
#plt.plot(df_Chile_inmigracion['agno'], df_Chile_inmigracion[col_3], label=label_col_3)
plt.plot(df_Chile_inclymin['agno'], df_Chile_inmigracion['Total'], label=label_total)

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title(f'PAÍS: CHILE\n{tema}')
plt.legend()
plt.grid(True)
plt.show()


#### Comentario a INMIGRACIÓN
* Este aspecto está definido por menciones en favor de la inmigración. Llama la atención el repunte fuerte en 2008 y luego la baja brusca en 2017, posiblemente por la asociación de la inmigración con la seguridad. 

#### MORAL TRADICIONAL (603, 604)

In [ ]:
# Define variables for the columns
tema = "MORAL TRADICIONAL"
col_1 = 'per603'
col_2 = 'per604'
#col_3 = 'per201_2'

# Select columns from df_Chile_ponderado
df_Chile_moral = df_Chile_ponderado[['agno', col_1, col_2]] #col_3

# Add a new column 'Total' with the sum of columns
df_Chile_moral['Total'] = df_Chile_moral[col_1] - df_Chile_moral[col_2] #+ df_Chile_moral[col_3]

# Get the column names excluding 'agno'
columns_to_lookup = df_Chile_moral.columns[df_Chile_moral.columns != 'agno']
detalle_dominio_dict = {}

# Loop through the columns and get the corresponding 'Detalle_Dominio' from 'df_AD'
for col in columns_to_lookup:
    detalle_dominio = df_AD.loc[df_AD['Variable_per'] == col, 'Detalle_Dominio'].values
    if len(detalle_dominio) > 0:
        detalle_dominio_dict[col] = detalle_dominio[0]

# Update the plot labels with details from the dictionary
label_col_1 = f"{col_1} ({detalle_dominio_dict.get(col_1, 'No Info')})"
label_col_2 = f"{col_2} ({detalle_dominio_dict.get(col_2, 'No Info')})"
#label_col_3 = f"{col_3} ({detalle_dominio_dict.get(col_3, 'No Info')})"
label_total = f"Total ({col_1} - {col_2})" # + {col_3})"

texto_md = """<div style="display: flex; justify-content: space-between;">
  <img src="../assets/per603_604.png" alt="" width="550"></div>
</div>
"""
display(Markdown(texto_md))

# Plot columns
plt.figure(figsize=(10, 6))
plt.plot(df_Chile_moral['agno'], df_Chile_moral[col_1], label=label_col_1)
plt.plot(df_Chile_moral['agno'], df_Chile_moral[col_2], label=label_col_2)
#plt.plot(df_Chile_moral['agno'], df_Chile_moral[col_3], label=label_col_3)
plt.plot(df_Chile_moral['agno'], df_Chile_moral['Total'], label=label_total)

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title(f'PAÍS: CHILE\n{tema}')
plt.legend()
plt.grid(True)
plt.show()


#### Comentario a MORAL TRADICIONAL: 
* Puede observarse como a partir de 2005 comienzan a ser mencionados los aspecto relacionados con el divorcio, el aborto y la concepción de familias con composición diferente a la tradicional en Chile.

#### GRÁFICO RESUMEN

In [ ]:
import matplotlib.pyplot as plt
plt.figure(figsize=(12, 8))

# Plot
plt.plot(df_Chile_poriginarios['agno'], df_Chile_poriginarios['Total'], label='Pueblos Originarios')
plt.plot(df_Chile_orden['agno'], df_Chile_orden['Total'], label='Orden y Corrupción')
plt.plot(df_Chile_medioambiente['agno'], df_Chile_medioambiente['Total'], label='Medio Ambiente y Sostenibilidad')
plt.plot(df_Chile_inclymin['agno'], df_Chile_inclymin['Total'], label='Inclusión y Minorías')
plt.plot(df_Chile_hr['agno'], df_Chile_hr['Total'], label='Derechos Humanos')
plt.plot(df_Chile_inmigracion['agno'], df_Chile_inmigracion['Total'], label='Inmigración')
plt.plot(df_Chile_moral['agno'], df_Chile_moral['Total'], label='Moral Tradicional')

# Etiquetas al final de cada línea
plt.annotate('Pueblos Originarios', xy=(df_Chile_poriginarios['agno'].values[-1], df_Chile_poriginarios['Total'].values[-1]), xytext=(5, 0), textcoords='offset points')
plt.annotate('Orden y Corrupción', xy=(df_Chile_orden['agno'].values[-1], df_Chile_orden['Total'].values[-1]), xytext=(5, 0), textcoords='offset points')
plt.annotate('Medio Ambiente y Sostenibilidad', xy=(df_Chile_medioambiente['agno'].values[-1], df_Chile_medioambiente['Total'].values[-1]), xytext=(5, 0), textcoords='offset points')
plt.annotate('Inclusión y Minorías', xy=(df_Chile_inclymin['agno'].values[-1], df_Chile_inclymin['Total'].values[-1]), xytext=(5, 0), textcoords='offset points')
plt.annotate('Derechos Humanos', xy=(df_Chile_hr['agno'].values[-1], df_Chile_hr['Total'].values[-1]), xytext=(5, 0), textcoords='offset points')
plt.annotate('Inmigración', xy=(df_Chile_inmigracion['agno'].values[-1], df_Chile_inmigracion['Total'].values[-1]), xytext=(5, 0), textcoords='offset points')
plt.annotate('Moral Tradicional', xy=(df_Chile_moral['agno'].values[-1], df_Chile_moral['Total'].values[-1]), xytext=(5, 0), textcoords='offset points')

# Configuración del gráfico
plt.xlabel('Año')
plt.ylabel('Total Ponderado')
plt.title('Comparación de Diferentes Aspectos en Chile')
plt.legend().set_visible(False)  # Oculta la leyenda ya que las etiquetas están al final de las líneas
plt.grid(True)

plt.show()


#### COMENTARIO GENERAL:
* Las conclusiones mencionadas en los gráficos individuales (resumidos en este cuadro comparativo), son consistentes con lo que se observa en Chile. Es posible que se pueda incorporar más variables a los aspectos mencionados de manera de tener una mejor representación, como también agregar más aspectos para analizar (Derechos Sociales, Economía, Sistema Político y otros).
* 
* En resumen, en base a este análisis, es posible sostener que el dataset habla de Chile como Chile es.

#### CHILE: TODOS LOS DOMINIOS

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Initialize an empty dataframe to store the results
df_Chile_dom = pd.DataFrame()

# Loop through each unique 'Dominio' in 'df_AD'
for dominio in df_AD['Dominio'].unique():
    # Create list 'columnas_per' with values of 'Variable_per' for that 'Dominio' considering only columns with name length == 6
    columnas_per = df_AD[(df_AD['Dominio'] == dominio) & (df_AD['Variable_per'].str.len() == 6)]['Variable_per'].tolist()
    
    # Sum the columns with names in 'columnas_per' for each 'agno' in 'df_Chile_ponderado'
    df_Chile_ponderado['tot_dom'] = df_Chile_ponderado[columnas_per].sum(axis=1)
    
    # Create a temporary dataframe with 'agno', 'Dominio', and 'tot_dom'
    temp_df = df_Chile_ponderado[['agno', 'tot_dom']].copy()
    temp_df['Dominio'] = dominio
    
    # Append the temporary dataframe to 'df_Chile_dom'
    df_Chile_dom = pd.concat([df_Chile_dom, temp_df], ignore_index=True)
    
    # Reset 'columnas_per' for the next loop iteration
    columnas_per = []

# Merge 'df_Chile_dom' with 'df_AD' to get 'Nombre_Dominio'
df_Chile_dom = df_Chile_dom.merge(df_AD[['Dominio', 'Nombre_Dominio']].drop_duplicates(), on='Dominio', how='left')

# Plot the dataframe 'df_Chile_dom' with 'agno' on the horizontal axis and 'tot_dom' on the vertical axis
plt.figure(figsize=(12, 8))

for dominio in df_Chile_dom['Dominio'].unique():
    subset = df_Chile_dom[df_Chile_dom['Dominio'] == dominio]
    
    # Plot the line for each 'Dominio'
    plt.plot(subset['agno'], subset['tot_dom'], label=subset['Nombre_Dominio'].iloc[0])
    
    # Check if 'Dominio' is 'Social Groups' or 'Freedom and Democracy' and place the label at the end
    if dominio in ['Domain_7', 'Domain_1']:
        plt.text(subset['agno'].iloc[-1], subset['tot_dom'].iloc[-1], subset['Nombre_Dominio'].iloc[0], 
                 fontsize=9, ha='left', va='center')
    else:
        # Place the label at the start for other dominios
        plt.text(subset['agno'].iloc[0], subset['tot_dom'].iloc[0], subset['Nombre_Dominio'].iloc[0], 
                 fontsize=9, ha='right', va='center')

# Configuración del gráfico
plt.xlabel('Año')
plt.ylabel('Total Dominio')
plt.title('Total Dominio por Año en Chile')
plt.legend()
plt.grid(True)
plt.show()


#### Alemania

In [ ]:
df_Alemania = df[df['countryname'] == 'Germany']

df_Alemania = df_Alemania.sort_values(by='agno')

# Create a list of column names that begin with 'per' and have a digit in position 4
per_columns_4digit = [col for col in df.columns if col.startswith('per') and col[3].isdigit()]

# Group by column 'agno' and calculate the weighted average of each column in 'per_columns_4digit' using column 'presvote' as weights
df_Alemania_ponderado = df_Alemania.groupby('agno').apply(
    lambda x: pd.Series({col: np.average(x[col], weights=x['pervote']) for col in per_columns_4digit})
).reset_index()

# Display the results
df_Alemania_ponderado.head()

In [ ]:
# Create dataframe
df_Alemania_inclymin = df_Alemania_ponderado[['agno', 'per503', 'per705']]

# Add a new column 'Total' with the sum of columns 'per503' and 'per705'
df_Alemania_inclymin['Total'] = df_Alemania_inclymin['per503'] + df_Alemania_inclymin['per705']

# Plot columns per503, per705, Total with agno in the horizontal axis
plt.figure(figsize=(10, 6))
plt.plot(df_Alemania_inclymin['agno'], df_Alemania_inclymin['per503'], label='per503')
plt.plot(df_Alemania_inclymin['agno'], df_Alemania_inclymin['per705'], label='per705')
plt.plot(df_Alemania_inclymin['agno'], df_Alemania_inclymin['Total'], label='Total')

plt.xlabel('Año')
plt.ylabel('Ponderado')
plt.title('PAÍS: ALEMANIA\nINCLUSIÓN Y MINORÍAS')
plt.legend()
plt.grid(True)
plt.show()

In [ ]:
plt.figure(figsize=(12, 8))

# Plot 'Total' for Chile
plt.plot(df_Chile_inclymin['agno'], df_Chile_inclymin['Total'], label='Chile Total')

# Plot 'Total' for Germany
plt.plot(df_Alemania_inclymin['agno'], df_Alemania_inclymin['Total'], label='Germany Total')

plt.xlabel('Año')
plt.ylabel('Total Ponderado')
plt.title('Comparación de Inclusión y Minorías: Chile vs Alemania')
plt.legend()
plt.grid(True)
plt.show()

### Diferencia entre Chile y Alemania

In [ ]:
# Merge the two dataframes on 'agno'
df_comparison = pd.merge(df_Chile_inclymin, df_Alemania_inclymin, on='agno', suffixes=('_Chile', '_Germany'))

# Calculate the difference between 'Total' for Chile and Germany
df_comparison['Total_Difference'] = df_comparison['Total_Chile'] - df_comparison['Total_Germany']

# Plot the results
plt.figure(figsize=(12, 8))
plt.plot(df_comparison['agno'], df_comparison['Total_Difference'], label='Difference in Total (Chile - Germany)')

plt.xlabel('Año')
plt.ylabel('Difference in Total Ponderado')
plt.title('Difference in Inclusión y Minorías: Chile vs Germany')
plt.legend()
plt.grid(True)
plt.show()

In [ ]:
df_comparison

#### Veamos la similaridad entre candidatos chilenos en suma de variables 503, 705, 201_2

In [ ]:
# Create list named 'variables'
variables = ['per503', 'per705', 'per201_2']

# Create dataframe 'df_Chile_serie_temporal' with columns 'agno', 'candidatename', and columns included in list 'variables'
df_Chile_serie_temporal = df_Chile[['agno', 'candidatename'] + variables]

# Create list 'candidatos_Chile' with unique values of column 'candidatename'
candidatos_Chile = df_Chile['candidatename'].unique()

# Create dataframe 'df_Chile_serie' with columns 'agno' and columns with the names in 'candidatos_Chile'
df_Chile_serie = pd.DataFrame({'agno': df_Chile['agno'].unique()})
for candidato in candidatos_Chile:
    df_Chile_serie[candidato] = np.nan

# Add column 'Total' to 'df_Chile_serie_temporal' with the sum of columns different from 'agno' and 'candidatename'
df_Chile_serie_temporal['Total'] = df_Chile_serie_temporal[variables].sum(axis=1)

# Fill 'df_Chile_serie' with values in column 'Total' of 'df_Chile_serie_temporal', for each 'agno' and 'candidatename'
for _, row in df_Chile_serie_temporal.iterrows():
    df_Chile_serie.loc[df_Chile_serie['agno'] == row['agno'], row['candidatename']] = row['Total']

df_Chile_serie

In [ ]:
plt.figure(figsize=(12, 8))

# Plot each candidate's 'Total' over the years
for candidato in df_Chile_serie.columns[1:]:
    plt.plot(df_Chile_serie['agno'], df_Chile_serie[candidato], label=candidato)

plt.xlabel('Año')
plt.ylabel('Total')
plt.title('Serie Temporal de Candidatos Chilenos\n Tema: Inclusión y Minorías')
plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
plt.grid(True)
plt.show()

#### Agreguemos todos los candidatos chilenos en las variables 503, 705, 201_2